### DISCUSSION TITLE
appender insert is very slow and the commit error is reported

### DISCUSSION THREAD

**Discussion by ethink** [duckdb/duckdb#3493] [General]
Hello
I wrote 50 million pieces of data with duckdb. Write speed is very slow.
In addition, JDBC transaction commit is not supported in duckdb.
Is there a good writing method for transaction submission?
The code is as follows:
#3489 
public static void  insertToFile() {
			 Connection conn = null;
			 Statement statement = null; 
			    ResultSet resultSet = null;
			    
			    try
			    {
			      Class.forName("com.mysql.jdbc.Driver");
			      conn = DriverManager.getConnection("jdbc:mysql://localhost:3306/api", "root", "root");
			      String sql="select * from duck5000wan";
			      long t1=System.currentTimeMillis();
			      PreparedStatement ps=(PreparedStatement)conn.prepareStatement(sql, ResultSet.TYPE_FORWARD_ONLY,ResultSet.CONCUR_READ_ONLY);
			      ps.setFetchSize(Integer.MIN_VALUE);
			      ps.setFetchDirection(ResultSet.FETCH_REVERSE);
			      resultSet= ps.executeQuery();
			      
			      
			      int index=0;
			      long t2=System.currentTimeMillis();
			      System.out.println("MySQL query time consuming："+(t2-t1));
			      Properties ro_prop = new Properties();
					ro_prop.setProperty("duckdb.read_only", "false");
					DuckDBConnection conn2 = (DuckDBConnection) DriverManager.getConnection("jdbc:duckdb:C:\\test\\duckdb\\premysql_test1_g_1.db", ro_prop);
					Statement stmt = conn2.createStatement();
					
					stmt.execute("CREATE TABLE data2 (s0 VARCHAR, s1 VARCHAR, s2 VARCHAR, s3 VARCHAR, s4 VARCHAR, s5 VARCHAR, s6 VARCHAR, s7 VARCHAR, s8 VARCHAR, s9 VARCHAR, s10 VARCHAR, s11 VARCHAR, s12 VARCHAR, s13 VARCHAR, s14 VARCHAR, s15 VARCHAR, s16 VARCHAR, s17 VARCHAR)");
					DuckDBAppender appender = conn2.createAppender("main", "data2");
				
			      while (resultSet.next())
			      {
			    	  index++;
			    	
			    	  appender.beginRow();
			    	  for(int i=1;i<=18;i++) {
			    		  Object oo = resultSet.getObject(4);
			    		  if(oo!=null) {
			    			  appender.append(oo.toString());
			    		  }
			    	  }
			    	  appender.endRow();
			      }
			      
			      long t3=System.currentTimeMillis();
			     
			      System.out.println("Duckdb insertion time："+(t3-t2));
			      System.out.println("Total time："+(t3-t1));
			    
			      stmt.close();
			      conn.close();
			    }catch(Exception e) {
			    	e.printStackTrace();
			    }

		}

**In addition, the errors reported during commit are as follows：**
	DuckDBAppender appender = conn2.createAppender("main", "data2");
				conn2.setAutoCommit(false);
			
		      while (resultSet.next())
		      {
		    	  index++;
		    	  appender.beginRow();	
		    	  for(int i=1;i<=18;i++) {
		    		  Object oo = resultSet.getObject(4);
		    		  if(oo!=null) {
		    			  appender.append(oo.toString());
		    		  }
		    		 
		    	  }
		    	  appender.endRow();
		       
		      conn2.commit();
**The background error log is as follows:**
java.sql.SQLException: TransactionContext Error: cannot commit - no transaction is active
	at org.duckdb.DuckDBNative.duckdb_jdbc_execute(Native Method)
	at org.duckdb.DuckDBPreparedStatement.execute(DuckDBPreparedStatement.java:89)
	at org.duckdb.DuckDBPreparedStatement.execute(DuckDBPreparedStatement.java:122)
	at org.duckdb.DuckDBConnection.commit(DuckDBConnection.java:80)

**Comment by SartorialOffense** [comment]
For inserting more than trivial amounts of data, you want to use the [Appender ](https://duckdb.org/docs/data/appender) (java usage at the bottom of https://duckdb.org/docs/data/overview) or import directly from Parquet or CSV. I typically just insert Parquet and it is very fast and easy.
